contracts/base/AaveV2Wrapper.sol

```
@@ -176,29 +253,89 @@ contract AaveV2Wrapper is OwnableUpgradeable, ReentrancyGuardUpgradeable {
         pool.withdraw(_asset, _amount, _to);
     }
 
+    /**
+     * @dev withdraw assets from AAVE and Treasury
+     * @param _asset asset address
+     * @param _amount asset amount
+     * @param _to receiving address
+     */
     function withdrawFromAave(
         address _asset,
         uint256 _amount,
         address _to
     ) internal {
         if (_amount == 0) revert ZeroAmount();
         if (_asset == address(0)) revert ZeroAddress();
-        if (lpTokens[_asset] == address(0)) revert AssetNotAllowed();
-        ILPtoken crToken = ILPtoken(lpTokens[_asset]);
-        if (_amount > crToken.balanceOf(msg.sender))
-            revert NotEnoughBalance();
+        if (crTokens[_asset] == address(0)) revert AssetNotAllowed();
+
+        ICRToken crToken = ICRToken(crTokens[_asset]);
+        if (_amount > crToken.balanceOf(msg.sender)) revert NotEnoughBalance();
         crToken.burn(msg.sender, _amount);
 
+        uint256 fromTreasury = _amount.mul(toTreasury).div(BASE);
+        uint256 fromAave = _amount.sub(fromTreasury);
+
         if (_asset == ETH) {
-            withdrawETH(_amount, _to);
+            withdrawETH(fromAave, _to);
         } else {
-            withdrawERC20(_asset, _amount, _to);
+            withdrawERC20(_asset, fromAave, _to);
         }
+
+        withdrawFromTreasury(_asset, _amount);
+        emit WithdrawEvent(_asset, _amount);
+    }
+
+    /**
+     * @dev Withdraw asset from treasury according to the user shares in the
+     * treasury.
+     * @param _asset asset address
+     * @param _amount asset amount
+     */
+    function withdrawFromTreasury(address _asset, uint256 _amount) internal {
+        uint256 current = balanceOf(_asset);
+        uint256 shareAmount = _amount.mul(current).div(pools[_asset]);
+        pools[_asset] = pools[_asset].sub(_amount);
+        if (_asset == ETH) {
+            require(payable(msg.sender).send(shareAmount));
```

we should return an error massage  in case transaction  fails